### PR TITLE
Add style guide for Go

### PR DESCRIPTION
**Pull request by infraredgirl**
This change adds a new category for Style guides and a style guide for Go under it.

Notes to reviewers:
- The category should be called "Style guides", not "Styleguides", but anchor links don't work properly if I do that. Maybe the platform team can help out with this one. 
- Totally open to using different icons, layout, section/category names etc. Let me know if you have any suggestions. I am not a designer! :smile: 
- Suggestions of what else to add to the guide content-wise are very welcome - since I am fairly new to the company, I am not yet familiar with all the particularities of our repos.

### CODE REVIEW

**Comment by zackproser** [comment]
This is awesome! Thanks for adding it - already learned from it.

File: _includes/search.html [hunk 1]
```diff
@@ -49,7 +49,8 @@
             <div class="center">
                 <div class="filter active-button" id="aws"><span class="helper"></span><img  src="/assets/img/logos-technologies/aws-logo.png" alt="aws logo"></div>
                 <div class="filter" id="gcp"><span class="helper"></span><img src="/assets/img/logos-technologies/gcp-logo.png" alt="aws logo"></div>
-                <div class="filter"id="azure"><span class="helper"></span><img src="/assets/img/logos-technologies/azure-logo.png" alt="aws logo"></div>
+                <div class="filter" id="azure"><span class="helper"></span><img src="/assets/img/logos-technologies/azure-logo.png" alt="aws logo"></div>
+                <div class="filter" id="styleguides"><span class="helper"></span><img src="/assets/img/logos-technologies/style-guides-logo.png" alt="style guides logo"></div>
```
Comment (yorinasub17) [inline _includes/search.html:53]: I'm not sure this fits in this filter, given that it has nothing to do with the cloud. I think what we actually want is the guide to be marked as all clouds, similar to https://github.com/gruntwork-io/gruntwork-io.github.io/blob/master/_posts/2019-08-26-how-to-use-gruntwork-infrastructure-as-code-library.adoc

File: _posts/2020-11-09-golang-style-guide.adoc [hunk 2]
```diff
@@ -0,0 +1,105 @@
+---
+title: Go Style Guide
+categories: Styleguides
+image: /assets/img/guides/go-logo-blue.png
+excerpt: The style guide for the Go programming language that is used by all Gruntwork-owned code repositories that contain Go code.
+tags: ["go", "golang", "code", "style"]
+cloud: ["styleguides"]
```
Comment (yorinasub17) [inline _posts/2020-11-09-golang-style-guide.adoc:7]: As mentioned above, I think we should actually mark this as all clouds to avoid overloading the cloud filter.

File: _posts/2020-11-09-golang-style-guide.adoc [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+Prefer using the `errors` standard library package for handling single errors and `hashicorp/go-multierror` for
+handling multiple errors.
+
+=== Pointer usage
+Prefer using value type over pointers, unless necessary.
+
+=== Testing
+All files containing business logic must have a corresponding unit test file, and we aim to have 100% test coverage.
+
+Unless there's a reason not to, tests should run in parallel. This is done by including a call to `t.Parallel` in the test function.
```
Comment (yorinasub17) [inline _posts/2020-11-09-golang-style-guide.adoc:86]: I think https://www.notion.so/gruntwork/Terraform-Module-Style-Guide-78084ab078e4472481b499a5767e136e#759fdeca732647c2a8c019dd1d9bf3c9 should be ported here.

**Review by yorinasub17 — COMMENTED** [review]
This LGTM! I had a few nits about the organization of the guides but other than that looks great!

**Review by zackproser — APPROVED** [review]
:tada:

File: _posts/2020-11-09-golang-style-guide.adoc (hunk 3 above)
Comment (robmorgan) [inline _posts/2020-11-09-golang-style-guide.adoc:86]:
```suggestion
Unless there's a reason not to, tests should run in parallel. This is done by including a call to `t.Parallel()` in the test function.
```
Should we add parenthesis to reinforce its a function call?

**Review by robmorgan — APPROVED** [review]

File: _posts/2020-11-09-golang-style-guide.adoc [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+=== Naming things
+Prefer descriptive names over short ones. In particular, avoid one-letter variable names.
```
Comment (brikis98) [inline _posts/2020-11-09-golang-style-guide.adoc:89]: NIT: mention that one-letter variable names are OK in a tiny handful of cases where that has become the idiom (e.g., `i` as the index in for-loops), but everywhere else, a more descriptive name helps with code understanding and maintenance, at very little cost, given auto-complete in most IDEs/editors.

File: _posts/2020-11-09-golang-style-guide.adoc [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+Since many languages use `ALL_CAPS` for constants, it is worth calling out explicitly that
+https://golang.org/doc/effective_go.html#mixed-caps[Effective Go] recommends using `MixedCaps` for all names, including constants.
+Therefore, `region` or `testRegion` for private constants and `Region` or `TestRegion` for public ones is preferred over
+`REGION` or `TEST_REGION`.
+
+=== Repo-specific conventions
+==== terratest
+Note the existence of methods in terratest which are suffixed with the letter `E`, e.g.
+https://github.com/gruntwork-io/terratest/blob/master/modules/aws/account.go#L23[GetAccountIdE]. The suffix `E`
+signifies that the method returns an error as the last return value.
```
Comment (brikis98) [inline _posts/2020-11-09-golang-style-guide.adoc:105]:
```suggestion
https://github.com/gruntwork-io/terratest/blob/master/modules/aws/account.go#L23[GetAccountIdE]. Methods that have the suffix `E` return an error as the last return value; methods without `E` mark the test as failed (e.g., via calling `t.Fail()`) instead of returning an error.
```

**Review by brikis98 — COMMENTED** [review]
Thanks for putting this together Ana! 

A few more items to add: 

1. **Default to functional programming practices.** That is:
    1. Immutability. Prefer returning a new value rather than mutating an existing one.
    1. Pure functions. Prefer functions that take all their inputs as function parameters, instead of reading those inputs via side effects (e.g., reading from disk or the network or global vars), and who's entire behavior is to return values (note: errors are values too!), rather than performing side effects (e.g., by writing to disk or the network or global vars). Of course, you can't avoid side effects forever if you want your code to do something useful, but try to do as much of your logic with pure functions as you can, and centralize the side effecting code to a few isolated places.
    1. Composition. Build your code out of small, named functions, that you compose together.
    1. Examples. Might be worth adding a few examples of the above.
1. **Error handling**. 
    1. Don't panic. Any method that can have an error should _return_ that error as its final value. This should be passed up through each layer, which can decide what to do with it, all the way up to the very entrypoint of the app (or `main`) if necessary. You should almost NEVER use `panic`.
    1. Use custom error types. Create your own types that implement the `error` interface so that error messages are clear and have well-defined types you can check against.
    1. Include stack traces. In most of our code, we have to wrap errors with `errors.WithStackTrace(e)` to add the stack trace. Go annoyingly doesn't do this by default, but without it, sorting out an error can be very tricky.
1. **Comments**. We may want to standardize on a comment style. E.g., Every public function `foo` has a `//` comment of the style `foo <explanation>`, where `explanation` says what `foo` does, and more importantly, why it does, what assumptions it makes, and other aspects not obvious from the function name.

File: _includes/search.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (infraredgirl) [reply _includes/search.html:53]: Totally agree wrt icon, however this issue is moot as I'll move the style guide under all clouds, as described [here](https://github.com/gruntwork-io/gruntwork-io.github.io/pull/373#discussion_r520145321).

File: _posts/2020-11-09-golang-style-guide.adoc (hunk 3 above)
Comment (infraredgirl) [reply _posts/2020-11-09-golang-style-guide.adoc:86]: I think it's fine, since we're talking about a function and referring to it by its name.

File: _posts/2020-11-09-golang-style-guide.adoc [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+== Starting point
+We use the excellent https://golang.org/doc/effective_go.html[Effective Go] guide as the starting point for our style
+guide. Unless explicitly mentioned otherwise, we default to following the conventions outlined in it.
+
+Another helpful resource is the https://github.com/golang/go/wiki/CodeReviewComments[CodeReviewComments] section of the
+Go GitHub wiki page.
+
+Below you will find the list of conventions that differ from the above mentioned documents, either because they are
+specific to Gruntwork, or because we consciously chose to go against what Effective Go and CodeReviewComments recommend.
```
[the remark this answers is not in the document]
Comment (yorinasub17) [reply _posts/2020-11-09-golang-style-guide.adoc:39]: I noticed this when doing the upgrade guide too. I didn't have time to figure out why this was happening, but I did manage to see that the CSS styles were only applying if you use the headings expected in the guides, `Intro`, `Core concepts`, `Deployment walkthrough`, etc. Just wanted to share this in case the info helps shorten the debug cycle.

File: _posts/2020-11-09-golang-style-guide.adoc [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+=== Error handling
+Prefer using the `errors` standard library package for handling single errors and `hashicorp/go-multierror` for
+handling multiple errors.
+
+=== Pointer usage
+Prefer using value type over pointers, unless necessary.
+
+=== Testing
+All files containing business logic must have a corresponding unit test file, and we aim to have 100% test coverage.
```
[the remark this answers is not in the document]
Comment (infraredgirl) [reply _posts/2020-11-09-golang-style-guide.adoc:84]: Makes sense, I'll reword.

**Comment by infraredgirl** [comment]
Addressed all the comments and added quite a bit more content and examples, so would appreciate another review.

Still not sure how to make the category section on the left behave correctly when I put two words there. It sould say "Style Guides" and not "Styleguides".
Screenshot from 2020-11-10 20-52-42

File: _posts/2020-11-09-golang-style-guide.adoc [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+} else {
+	if orOther {
+		return that
+	}
+}
+----
+
+=== Error handling
+Prefer using the `errors` standard library package for handling single errors and `hashicorp/go-multierror` for
+handling multiple errors.
```
[the remark this answers is not in the document]
Comment (yorinasub17) [reply _posts/2020-11-09-golang-style-guide.adoc:82]: NOTE: We want to avoid linking in hashicorp libraries as they use a flavor of copyleft license (albeit weak form) which we currently don't allow use of as a policy. Can't do anything about the usage we already have, but should probably avoid new usage until we resolve that issue.

Given that, we probably shouldn't include recommendations in the style guide that require linking in a hashicorp library.

File: _posts/2020-11-09-golang-style-guide.adoc [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+==== Constants
+Since many languages use `ALL_CAPS` for constants, it is worth calling out explicitly that
+https://golang.org/doc/effective_go.html#mixed-caps[Effective Go] recommends using `MixedCaps` for all names, including constants.
+Therefore, `region` or `testRegion` for private constants and `Region` or `TestRegion` for public ones is preferred over
+`REGION` or `TEST_REGION`.
+
+=== Repo-specific conventions
+==== terratest
+Note the existence of methods in terratest which are suffixed with the letter `E`, e.g.
```
[the remark this answers is not in the document]
Comment (yorinasub17) [reply _posts/2020-11-09-golang-style-guide.adoc:294]: I don't think so, but we should add as we find them.

**Review by yorinasub17 — COMMENTED** [review]
Updates LGTM! The one thing that I think we should tweak is the section on errors, where we reference a hashicorp library.

File: _posts/2020-11-09-golang-style-guide.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (infraredgirl) [reply _posts/2020-11-09-golang-style-guide.adoc:82]: Ah, that's a good point that totally slipped my mind. Will amend.

**Review by brikis98 — APPROVED** [review]
LGTM! @eak12913 and @oredavids we'd still appreciate your help on the paragraph spacing issue.

**Comment by infraredgirl** [comment]
Thanks folks for the reviews!

I'll hold off on merging until we get some input on the paragraph spacing and the category naming issues.

**Comment by bwhaley** [comment]
This is fantastic! Nice work. The only feedback I have is to consider mentioning the [gruntwork-cli](https://github.com/gruntwork-io/gruntwork-cli) repo which contains a bunch of packages for Go CLI programs, including errors, logging, shell helpers, etc.

**Comment by eak12913** [comment]
Ugh - I hate _hate_ github notifications. I apologize to all for seeing this so late. Could I ask for a personal favor, if folks need my attention - could you mention me in our Slack channel. This PR has many messages and the specific call out to me got lost in my inbox :-(

**Comment by eak12913** [comment]
> LGTM! @eak12913 and @oredavids we'd still appreciate your help on the paragraph spacing issue.

@brikis98 @infraredgirl --- 2 1/2 hours later I found the issue. The problem is with the code in `search.js`. 

TLDR; one part of the code is comparing a kebab cased identifier vs a non kebab cased identifier. Those are identical when there are no spaces, but when spaces are introduced - the comparison fails because you end up with `style-guide` being compared to `style guide` and then the section is just not being shown.

I'm going to bed now but I'll push the fix tomorrow!

**Comment by infraredgirl** [comment]
Thanks Eugene, and no worries! GitHub notifications are far from perfect (and this was by no means urgent anyway)!

**Comment by infraredgirl** [comment]
test @eak12913

**Comment by eak12913** [comment]
@brikis98 @infraredgirl I pushed my changes to fix the issue. Do you folks want to take a look and approve/merge away as necessary?

**Comment by infraredgirl** [comment]
Hey @eak12913 that looks much better, thanks for your help! The only problem I still see the link in the breadcrumbs is not rendering nor working properly:

Screenshot from 2020-11-17 17-12-35

Any chance you could help with that also? :smile:

**Comment by eak12913** [comment]
> Hey @eak12913 that looks much better, thanks for your help! The only problem I still see the link in the breadcrumbs is not rendering nor working properly:
> Any chance you could help with that also? 😄

I believe I've fixed this. Please take a look!

**Comment by eak12913** [comment]
The only bit of slight annoyance is that we're only capitalizing the first word...so it's: "Style guides" in the breadcrumb and not "Style Guides". I'm not sure if this is a problem.  It's a little more annoying to fix this with Jekyll - but doable so if it's important I can go back in there and figure it out.

Additionally it should also be possible to do this with CSS - but then that may change the capitalization of a title where we intentionally have "The first thing capitalized but nothing else" into --> "The First Thing Capitalized But Nothing Else"

**Comment by infraredgirl** [comment]
Looks perfect, many thanks @eak12913! (I can definitely live with "Style guides" vs "Style Guides".)

@gruntwork-io/team-iac can one of you give the final seal of approval, please?

**Review by brikis98 — APPROVED** [review]
LGTM!

**Comment by brikis98** [comment]
BTW, we'll probably want to cross-link to this from our internal Notion pages!

And eventually, migrate our Terraform style guide here...

**Comment by infraredgirl** [comment]
> BTW, we'll probably want to cross-link to this from our internal Notion pages!
> 
> And eventually, migrate our Terraform style guide here...

Definitely - I'll take care of it. 

We should also link to this guide from our open source repos that are heavy on Go (terragrunt, terratest, cloud-nuke, any others?) - I'll do that too.

**Comment by brikis98** [comment]
> Definitely - I'll take care of it.

Thanks! It's not urgent, so feel free to tackle it incrementally, when time permits.

> We should also link to this guide from our open source repos that are heavy on Go (terragrunt, terratest, cloud-nuke, any others?) - I'll do that too.

Yup, good idea!

File: _posts/2020-11-09-golang-style-guide.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (brikis98) [reply _posts/2020-11-09-golang-style-guide.adoc]: @eak12913 I noticed this issue still remains:

Screen Shot 2020-11-18 at 12 58 01 PM

Note the lack of vertical whitespace between paragraphs. Not urgent, but when you get a chance, could you take a look?

[the remark this answers is not in the document]
Comment (brikis98) [reply _posts/2020-11-09-golang-style-guide.adoc]: @eak12913 -- test mention as requested.

[the remark this answers is not in the document]
Comment (eak12913) [reply _posts/2020-11-09-golang-style-guide.adoc]: Thanks!

PR to address the issue here: https://github.com/gruntwork-io/gruntwork-io.github.io/pull/379
